### DISCUSSION TITLE
Importing spice models to qucs user library

### DISCUSSION THREAD

**Discussion by bnilsson11** [ra3xdh/qucs_s#792] [Q&A] [answered]
I want to import the following simple spice model ty my user library:

~~~
*model J505	anode
*		| cathode
*		| |
.subckt J505	1 2
  j1 1 2 2 jx
  .model   jx	NJF(Beta=100u Betatce=-.5 Rd=1 Rs=0 Lambda=526u Vto=-3.158
+		Vtotc=-2.5m N=2 Xti=3 Cgd=2p Is=10f M=.5 Pb=1 Fc=.5 Af=1)
*		Siliconix	pid=NCL		case=TO92mod
*		88-07-22 bam	POVmax=50
*		88-12-16 pwt
.ends
~~~

Is there any automated method to do this, och should I create <Component>....</Component> manually by studying existing library entries? I tried qucsconv, but it was not helpful.
Anything in the new docs about it?

BN

**Comment by ra3xdh** [comment, marked as answer]
You have to wrap the model in `SpiceFile` device and subcircuit. See this article: https://habr.com/ru/articles/805963/ Sorry, there is no english version, but it is possible to understand what is happening from the screenshots in the article. The newest version has introduced a special *Spice Library Device*. See #679, #775, and #781 for details. No sibcircuit wrapper required anymore. You may try to build Qucs-S from Git and use the new device.

**Comment by bnilsson11** [comment]
Thanks a lot.
I would be happy to build the latest version, I have 24.2.1 now.
(A comment: "About qucs-s" on my MacOS platform shows ONLY the mandatory Qt information, nothing about qucs-s itself. I have to know myself which version I have built...)
Your build instructions for ubuntu does not apply to my platform in the details.
What is the "git clone ..." command line to get the latest version?

**Reply by ra3xdh** [reply]
I cannot suggest anything about how to build Qucs-S on Mac platform, because I have no access to Mac hardware. You may refer to CI build script  https://github.com/ra3xdh/qucs_s/blob/current/.github/workflows/macos.yml and #181 to find some info. The `git` is a version control system that clones the source repository. The `git` tool must be installed on your OS. Probably it may be better to wait the next release in your case.

**Comment by bnilsson11** [comment]
I think you misunderstand me.
I simply need instructions how to get the latest tarball.
The rest of the build I have done before.



> 2 juli 2024 kl. 13:37 skrev Vadim Kuznetsov ***@***.***>:
> 
> 
> I cannot suggest anything about how to build Qucs-S on Mac platform, because I have no access to Mac hardware. You may refer to CI build script https://github.com/ra3xdh/qucs_s/blob/current/.github/workflows/macos.yml and #181 <https://github.com/ra3xdh/qucs_s/discussions/181> to find some info. The git is a version control system that clones the source repository. The git tool must be installed on your OS. Probably it may be better to wait the next release in your case.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ra3xdh/qucs_s/discussions/792#discussioncomment-9935722>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAZXLDXPKKJUOV43NT2SXLDZKKGHZAVCNFSM6AAAAABKHJVVTWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TSMZVG4ZDE>.
> You are receiving this because you authored the thread.
> 

Bengt Nilsson
Ansåsvägen 8
43853 Hindås
Sweden
***@***.***

**Reply by ra3xdh** [reply]
If `git` is not installed on your system, you may download the tarball from the continuous release page: https://github.com/ra3xdh/qucs_s/releases/tag/continuous_build It contains three archive: windows binary, zip and tar with source.

**Comment by bnilsson11** [comment]
Ok, back on track with the SPICE model import to qucs-s.
I tried the SPICE Library Device with the J505 model shown in the beginning and it worked for me at my first attempt, fantastic. 
A very simple model, but still.

The only issue was that I could not find any suitable symbol template.
I have to admit I do not know to make a suitable symbol file.
What should be the content, extension?
File/New Symbol seems not correct for this?

**Comment by ra3xdh** [comment]
> I have to admit I do not know to make a suitable symbol file.

I will upload a diode symbol pattern in the next release. Use *File->New symbol* to draw a symbol. Then use *Symbol from file* and specify the user file. The symbol file has the `*.sym` extension.

**Comment by bnilsson11** [comment]
Tried it, and it did not work as expected.
Project is enclosed below.
The .sym file has the graphical symbol, two port handles 1 and 2, and the SUB part.
Initially, I used the `diac` symbol, tested a simulation and it was working well as far as I could see.
I opened my SPICE Library Device and changed from symbol template to file symbol with my .sym file.
Applied and Closed, then the symbol became in the schematic what I think is the automatic symbol with ports direction horizontally.
I double clicked on it to open it again and **then qucs-s crashed**.
I repeated the procedure, rotated the automatic box symbol device and ran the simulation and the sim was ok.
So the SPICE Library Device gets the wrong symbol (for me) but the sim works ok, and it cannot be edited since qucs-s crashes..
[Constant_current_source_prj.zip](https://github.com/user-attachments/files/16097966/Constant_current_source_prj.zip)

**Comment by ra3xdh** [comment]
I have fixed the reported problems with user symbol for SpiceLibComp by #802 Please update to the latest continuous release.

**Comment by bnilsson11** [comment]
Thanks, it is now working.
I found another connected issue in the SPICE Library Device maybe worth looking into.
When changing the pin configuration, the "remaining pins" logic do not release the available pins as it should.
If you do a mistake and want to reconfigure, you sometimes cannot, the pins you want to select are not available.
if you set all to NC you should have the full list to select from, but you get maybe only one (usually the one you want to change from)  or none.
Closing and reopen the dialog does not help.

I am fully aware the I am sometimes hard to understand when I try to explain things but I hope you can follow my description.

**Comment by ra3xdh** [comment]
> the "remaining pins" logic do not release the available pins as it should.

It is not a bug. Just clear assignment (assign NC pin) and then reassign the pin. This is made to prevent the duplicate pin assignment.

**Comment by bnilsson11** [comment]
Not working.As I wrote, assigning all pins to NC initially do not make all pins available in the new assignment attempt.MvhBengt NilssonSkickat från min iPhone4 juli 2024 kl. 18:52 skrev Vadim Kuznetsov ***@***.***>:﻿

the "remaining pins" logic do not release the available pins as it should.

It is not a bug. Just clear assignment (assign NC pin) and then reassign the pin. This is made to prevent the duplicate pin assignment.

**Comment by ra3xdh** [comment]
Everything works as expected on my machine. See the attached video. The pins are released. 

[Screencast_20240704_200738.webm](https://github.com/ra3xdh/qucs_s/assets/4920080/3a6c1557-8551-49b7-8b4c-ee6eaa9f3f3d)

**Comment by bnilsson11** [comment]
Looks fine at your site, maybe it is a macos qt issue, it has happened before. Or something else.I will send you a project where I have the problem.MvhBengt NilssonSkickat från min iPhone4 juli 2024 kl. 19:10 skrev Vadim Kuznetsov ***@***.***>:﻿
Everything works as expected on my machine. See the attached video. The pins are released.
Screencast_20240704_200738.webm

**Comment by bnilsson11** [comment]
Looks fine at your site, maybe it is a macos qt issue, it has happened before. Or something else.
I send you a project where I have the problem, and a screencast.

Please note that this sim is not working properly, I don't know why.
It is a 3-pin temperature sensor LM335, 10mV/K, oddly you set the sim temperature by a fourth pin.
27 V on the T pin should give a voltage of around 3.00V (300K).
If you can figure out quickly why it is not working, it would be interesting to know. If not, drop it, it is not the issue here.
It is difficult to see how this can affect the pin config behaviour.


Mvh
Bengt Nilsson
[Temp_sensor_prj.zip](https://github.com/user-attachments/files/16102522/Temp_sensor_prj.zip)

**Comment by ra3xdh** [comment]
> Looks fine at your site, maybe it is a macos qt issue

Yes, it seems to be a macos issue. The selection of the table cells works differently on MacOS. 

> Please note that this sim is not working properly

You have an obvious error in schematic. The anode and cathode of the LM335 are flipped. The simulation works correctly after flipping the device vertically.
![image](https://github.com/ra3xdh/qucs_s/assets/4920080/5b587f33-2d39-409a-b162-889bed26380e)

**Comment by bnilsson11** [comment]
Thanks, the obvious errors are sometimes most difficult to see.
If I had drawn the symbol as a zener I probably wouldn’t have made this mistake.

Good that the cell selection problem is identified.


> 5 juli 2024 kl. 08:18 skrev Vadim Kuznetsov ***@***.***>:
> 
> 
> Looks fine at your site, maybe it is a macos qt issue
> 
> Yes, it seems to be a macos issue. The selection of the table cells works differently on MacOS.
> 
> Please note that this sim is not working properly
> 
> You have an obvious error in schematic. The anode and cathode of the LM335 are flipped. The simulation works correctly after flipping the device vertically.
> image.png (view on web) <https://github.com/ra3xdh/qucs_s/assets/4920080/5b587f33-2d39-409a-b162-889bed26380e>
> —
> Reply to this email directly, view it on GitHub <https://github.com/ra3xdh/qucs_s/discussions/792#discussioncomment-9964774>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAZXLDQ26OSMLDJQEINFREDZKY3C3AVCNFSM6AAAAABKHJVVTWVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TSNRUG43TI>.
> You are receiving this because you authored the thread.
> 

Bengt Nilsson
Ansåsvägen 8
43853 Hindås
Sweden
***@***.***

**Comment by bnilsson11** [comment]
I saw a new release today, tried it and there was no improvement in the menu cell handling. 
Is this issue planned for later?

**Reply by ra3xdh** [reply]
The pin assignment issue has been fixed by #824

**Comment by ra3xdh** [comment]
The latest snapshot doesn't contain an update related to this issue. I have created #808. It's need to wait for developer using MacOS to fix this. I am unable to provide fix myself.

**Comment by evenicholson12** [comment]
This is a useful discussion for anyone working with SPICE models in Qucs-S. The step-by-step troubleshooting around model imports, custom symbols, pin configuration, and platform-specific behaviour shows how practical community support can make technical projects much easier to manage. The Academic Papers UK also provides **[Assignment Writing Services](https://www.theacademicpapers.co.uk/assignment-writing-services/
)** support for students working on technical and research-based coursework.
